File: problems/16.graphs/ts/graph/Graph.ts
export interface Graph<T> {
    vertices: T[];
    edges: number[][];
    addVertex(vertex: T): Graph<T>;
    addEdge(v: T, w: T): Graph<T>;
}

export class GraphImpl<T extends number> implements Graph<T> {
    vertices: T[];
    edges: number[][];

    constructor() {
        this.vertices = [];
        this.edges = [];
    }

    public addVertex(vertex: T) {
        this.vertices.push(vertex);
        this.edges[vertex] = [];

        return this;
    }

    public addEdge(v: T, w: T) {
        this.edges[v].push(w);
        this.edges[w].push(v);
        return this;
    }

    public dfsTraverse(
        vertex: T,
        adjList: T[][],
        result: T[],
        seen: Map<T, boolean>
    ) {
        if (vertex === undefined || seen.get(vertex)) return result;
        result.push(vertex);

        seen.set(vertex, true);
        let connections = adjList[vertex];

        for (let i = 0; i < connections.length; i++) {
            let connection = connections[i];
            if (!seen.get(connection)) {
                this.dfsTraverse(connection, adjList, result, seen);
            }
        }
    }

    public bfsTraverse(): number[] {
        let adjList = this.edges;
        let queue = [0];
        let seen = new Map<number, boolean>();
        const result = [];

        while (queue.length > 0) {
            let vertex = queue.shift() as number;
            seen.set(vertex, true);
            result.push(vertex);
            let connections = adjList[vertex];

            for (let i = 0; i < connections.length; i++) {
                let node = connections[i];
                if (!seen.get(node)) {
                    queue.push(node);
                }
            }
        }

        return result;
    }
}

File: problems/16.graphs/golang/graph/graph.go
package graph

import "fmt"

type Graph struct {
	vertices []int
	edges    [][]int
}

func NewGraph() Graph {
	return Graph{
		vertices: []int{},
		edges:    [][]int{},
	}
}

func (g Graph) AddVertex(vertex int) Graph {
	g.vertices = append(g.vertices, vertex)

	if vertex < len(g.edges) {
		g.edges[vertex] = []int{}
	} else {
		g.edges = make([][]int, vertex+1)
	}

	return g
}

func (g Graph) AddEdge(from, to int) Graph {
	g.edges[from] = append(g.edges[from], to)
	g.edges[to] = append(g.edges[to], from)

	return g
}

func (g Graph) BfsTraverse() []int {
	queue := []int{0}
	result := []int{0}
	seen := make(map[int]bool)
	seen[0] = true
	edges := g.edges

	for len(queue) > 0 {
		vertex := queue[0]
		queue = queue[1:]
		connections := edges[vertex]
		for i := 0; i < len(connections); i++ {
			connection := connections[i]
			if !seen[connection] {
				queue = append(queue, connection)
				result = append(result, connection)
			}
			seen[connection] = true
		}
	}

	return result
}

func dfs(edges [][]int, seen map[int]bool, vertex int, result *[]int) {
	// kill condition
	if seen[vertex] {
		return
	}

	// add to result
	seen[vertex] = true
	*result = append(*result, vertex)
	connections := edges[vertex]

	// call dfs again
	for _, connection := range connections {
		if !seen[connection] {
			dfs(edges, seen, connection, result)
		}
	}
}

func (g Graph) DfsTraverse() []int {
	seen := make(map[int]bool, len(g.vertices))
	result := []int{}
	dfs(g.edges, seen, 0, &result)

	return result
}
